### PR TITLE
add interactive VMs

### PR DESCRIPTION
**Pull request by Lassulus**

### CODE REVIEW

File: lib/types/luks.nix [hunk 1]
```diff
@@ -114,19 +114,23 @@ in
...
           ${lib.optionalString config.askPassword ''
-            set +x
             askPassword() {
-              echo "Enter password for ${config.device}: "
-              IFS= read -r -s password
-              echo "Enter password for ${config.device} again to be safe: "
-              IFS= read -r -s password_check
-              export password
-              [ "$password" = "$password_check" ]
+              if [ -z ''${IN_DISKO_TEST+x} ]; then
```
Comment (phaer) [inline lib/types/luks.nix:118]: Why is this needed? To an interactive password code path as opposed to just key files?

I think it can be a bit smelly to let code check whether it's in a test here 😅

File: docs/interactive-vm.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Running Interactive VMs with disko
+
+disko now exports it's own flavor of interactive VMs (similiar to config.system.build.vm).
```
Comment (Mic92) [inline docs/interactive-vm.md:3]: Since we normal build-vm stuff is broken, should we maybe mention this here?

File: docs/interactive-vm.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+Simply import the disko module and build the vm runner with:
+```
+nix build -L '.#nixosConfigurations.mymachine.config.system.build.vmWithDisko'
+```
+
+afterwards you can run the interactive VM with:
+
+```
+result/bin/disko-vm
```
Comment (Mic92) [inline docs/interactive-vm.md:12]: Could potentially become part of the disko cli in another PR.

File: lib/interactive-vm.nix [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+in
+{
+  pure = pkgs.writeDashBin "disko-vm" ''
+    set -efux
+    export tmp=$(mktemp -d)
+    trap 'rm -rf "$tmp"' EXIT
+    ${lib.concatMapStringsSep "\n" (disk: ''
+      ${pkgs.qemu}/bin/qemu-img create -f qcow2 \
+      -b ${diskoImages}/${disk.name}.raw \
+      -F raw "$tmp"/${disk.name}.qcow2
```
Comment (Mic92) [inline lib/interactive-vm.nix:83]: How does this cope with different disk sizes? Also why are not using qcow here?

File: lib/types/luks.nix (hunk 1 above)
Comment (Mic92) [reply lib/types/luks.nix:118]: Can we not just override passwordFile in disko?

File: lib/make-disk-image.nix [hunk 5]
```diff
@@ -15,10 +18,14 @@ let
...
   cleanedConfig = diskoLib.testLib.prepareDiskoConfig nixosConfig.config diskoLib.testLib.devices;
   systemToInstall = nixosConfig.extendModules {
-    modules = [{
-      disko.devices = lib.mkForce cleanedConfig.disko.devices;
-      boot.loader.grub.devices = lib.mkForce cleanedConfig.boot.loader.grub.devices;
-    }];
+    modules = [
+      extraConfig
+      {
+        disko.testMode = true;
```
Comment (Mic92) [inline lib/make-disk-image.nix:24]: We don't need testMode if we can override the password file in luks.

File: lib/types/luks.nix (hunk 1 above)
Comment (Lassulus) [reply lib/types/luks.nix:118]: no, that's not possible since disko doesn't merge the devices correctly

File: lib/make-disk-image.nix (hunk 5 above)
Comment (Lassulus) [reply lib/make-disk-image.nix:24]: password cannot be overwritten, so testMode is the easiest way to fix that

File: lib/interactive-vm.nix (hunk 4 above)
Comment (Lassulus) [reply lib/interactive-vm.nix:83]: this creates a COW overlay image, with the backing one the raw image generated by the imagebuilder

File: lib/types/luks.nix (hunk 1 above)
Comment (Mic92) [reply lib/types/luks.nix:118]: Can we set a "default" with higher than normal priority through "options" instead?

File: lib/interactive-vm.nix (hunk 4 above)
Comment (Mic92) [reply lib/interactive-vm.nix:83]: It would be good if we also could have qcow on the disko images side later. Because than we can make these images sparse and not explode in size.

Comment (Lassulus) [reply lib/interactive-vm.nix:83]: thats possible with the extraPostVM hook already, but yeah maybe it would be better to use that here?

File: lib/types/luks.nix (hunk 1 above)
Comment (Lassulus) [reply lib/types/luks.nix:118]: no, since we can't modify the the options after we evaluate the vm attribute, because the config is already evaluated

File: docs/interactive-vm.md (hunk 2 above)
Comment (Lassulus) [reply docs/interactive-vm.md:3]: whats broken?

Comment (Mic92) [reply docs/interactive-vm.md:3]: See the details here: https://github.com/nix-community/disko/issues/668
